server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.dml;
+
+import org.jetbrains.annotations.Nullable;
+
+import io.crate.data.Row1;
+
+public final class RowCountAndFailure {
```

Comment:
I know PR is still in draft but drive-by comment anyway:

Maybe worth to go with a struct of arrays instead of array of structs here and have something like `Rowcounts` / `BulkResponse` that has `long[] counts` and `Throwable[] failures` for a smaller memory footprint.

And I think we should look into limiting the amount of `Throwable` instances we keep - e.g. maybe only keep up to 50 or 100 errors, and discard afterwards.
E.g. with a bulk size of 10_000 and each item failiing, and each having a stack of somewhere of 50-100 we could end up using lots of memory and running into circuit breaker/OOM because of the failure tracking.

Looks like `Throwable` is used for the permission checks - but maybe we could even do these eagerly and only keep the message as string?

Reply:
Thanks for the feedback, valid points, will look into them.

> Looks like Throwable is used for the permission checks - but maybe we could even do these eagerly and only keep the message as string?

Not only permission checks, but also to convert them into user friendly exceptions.
In order to move the permission check on the shard, we'd need to stream the `CoordinatorSessionSettings` (so they wouldn't be *Coordinator*  once anymore). But we probably need to do this anyhow, to fix information leaks already existing when using `RETURN SUMMARY`, see https://github.com/crate/crate/pull/16799/files#diff-ccd3239453fb803999cc796e8f97365611ff44ead633df9db0acdeda147bb3b4R433.

Reply:
Another issue with converting the `Throwable` directly into a message, is the error `code`. This code is very interface related, we currently generate it out of the `Throwable` only on the REST interface. The PG interface may create different error codes. I see these options when we want to avoid streaming the exception:
- a) Create the error code based on the error message   <- sounds very error-prone
- b) Create the error codes at the shards and transport them <- moves interface(API) related code deep into the execution
- c) Create internal codes we can use to create API specific codes
- d) ?

I'd favor c).

Reply:
For c) we may could reuse the already existing exception id's of `ElasticsearchExceptionHandle`.

Reply:
I've pushed a rough draft implementation of c) by https://github.com/crate/crate/pull/16799/commits/0ca94ee0d9b6e4f003489c864003a5a07a7185ef, please also check this commit comments.

Reply:
I'm not sure either about which way to go but I'd almost tend to keep the exceptions but ensure we never keep more than ~50 or so in memory.
And also change the existing ShardResponse.Failure to work with exceptions instead of messages.

The main reason is that converting exceptions just before they're sent to the client to account for privileges seems like an easier model to reason about than if the logic gets done in different phases.

And keeping the original exceptions around for longer might help with debugging during development.